gc/base/EvacuatorBase.hpp

```diff
@@ -0,0 +1,214 @@
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef EVACUATORBASE_HPP_
+#define EVACUATORBASE_HPP_
+
+#undef EVACUATOR_DEBUG
+#undef EVACUATOR_DEBUG_ALWAYS
+
+#if defined(EVACUATOR_DEBUG)
+#include "omrgcconsts.h"
+#include "ModronAssertions.h"
+
+#include "Math.hpp"
+
+/* base debug flags */
+#define EVACUATOR_DEBUG_END 1
+#define EVACUATOR_DEBUG_CYCLE 2
+#define EVACUATOR_DEBUG_EPOCH 4
+#define EVACUATOR_DEBUG_WORK 8
+#define EVACUATOR_DEBUG_STACK 16
+#define EVACUATOR_DEBUG_COPY 32
+#define EVACUATOR_DEBUG_ALLOCATE 64
+#define EVACUATOR_DEBUG_WHITELISTS 128
+#define EVACUATOR_DEBUG_POISON_DISCARD 256
+#define EVACUATOR_DEBUG_BACKOUT 512
+#define EVACUATOR_DEBUG_DELEGATE 1024
+
+/* default debug flags */
+#define EVACUATOR_DEBUG_DEFAULT_FLAGS (0)
+
+/* delegate can define additional flags above 0x10000 */
+#define EVACUATOR_DEBUG_DELEGATE_BASE 0x10000
+
+#define Debug_MM_true(assertion) Assert_MM_true(assertion)
+#define Debug_MM_true1(env, assertion, format, arg) Assert_GC_true_with_message1(env, assertion, format, arg)
+#define Debug_MM_true2(env, assertion, format, arg1, arg2) Assert_GC_true_with_message2(env, assertion, format, arg1, arg2)
+#define Debug_MM_true3(env, assertion, format, arg1, arg2, arg3) Assert_GC_true_with_message3(env, assertion, format, arg1, arg2, arg3)
+#define Debug_MM_true4(env, assertion, format, arg1, arg2, arg3, arg4) Assert_GC_true_with_message4(env, assertion, format, arg1, arg2, arg3, arg4)
+#else
+#define Debug_MM_true(assertion)
+#define Debug_MM_true1(env, assertion, format, arg)
+#define Debug_MM_true2(env, assertion, format, arg1, arg2)
+#define Debug_MM_true3(env, assertion, format, arg1, arg2, arg3)
+#define Debug_MM_true4(env, assertion, format, arg1, arg2, arg3, arg4)
+#endif /* defined(EVACUATOR_DEBUG) */
+
+#include "GCExtensionsBase.hpp"
+
+#if (DEFAULT_SCAN_CACHE_MINIMUM_SIZE > (DEFAULT_SCAN_CACHE_MAXIMUM_SIZE >> 3))
+#error "Scan cache default sizes must satisfy DEFAULT_SCAN_CACHE_MINIMUM_SIZE <= (DEFAULT_SCAN_CACHE_MAXIMUM_SIZE >> 3)"
```

Comment:
Can you explain this more with a comment?

Reply:
Will do. Evacuator wants to ensure that there is a reasonable range between min/max whitespace allocation sizes.